File: src/main/java/org/alham/alhamfirst/service/orchestrator/todo/TodoServiceImpl.java
package org.alham.alhamfirst.service.orchestrator.todo;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.alham.alhamfirst.dto.todo.TodoDTO;
import org.alham.alhamfirst.entity.User;
import org.alham.alhamfirst.entity.todo.Todo;
import org.alham.alhamfirst.mapper.TodoMapper;
import org.alham.alhamfirst.repository.todo.TodoRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class TodoServiceImpl implements TodoService{

    private final TodoRepository todoRepository;

    private final TodoMapper todoMapper;

    @Override
    public TodoDTO getTodoDetailByUserId(Long id) {
        Todo byUserId = todoRepository.findByUserId(id);
        return todoMapper.createTodoDTOFromEntity(byUserId);
    }

    /**
     * 할일 생성
     * startDate는 해당 레이어에서 넣어줌
     *
     * @param todoDTO
     * @return
     */
    @Override
    @Transactional
    public Todo createTodo(TodoDTO todoDTO) {
        todoDTO.setStartDate(LocalDate.now());
        Todo todo = todoMapper.createTodoFromDTO(todoDTO);
        User user = User.createTempUser(todoDTO.getUserId());
        todo.addUser(user);
        return todoRepository.save(todo);
    }

    @Override
    public List<TodoDTO> getTodoList() {
        return todoRepository.findAll().stream().map(todoMapper::createTodoDTOFromEntity).toList();
    }

    @Override
    public List<TodoDTO> listTodo() {
        return todoRepository.findAll().stream().map(todoMapper::createTodoDTOFromEntity).toList();
    }

    @Override
    public TodoDTO getTodoDetail(Long id) {
        return todoRepository.findById(id).map(todoMapper::createTodoDTOFromEntity).orElse(new TodoDTO());
    }

    //TODO - 업데이트 할때마다 stat 변경 로직 적용여부 확인 후 적용
    @Override
    @Transactional
    public TodoDTO updateTodoDetail(TodoDTO todoDTO) {
        Todo todo = todoRepository.findById(todoDTO.getId()).orElseThrow();
        todo.updateTodo(todoDTO.getDetail());
        Todo saved = todoRepository.save(todo);
        return todoMapper.createTodoDTOWithUserIdFromEntity(saved);

    }

    @Override
    public List<TodoDTO> getTodoListByUserIdWithUndo(Long id) {
        List<Todo> undoListByUserId = todoRepository.findUndoListByUserId(id);
        return todoRepository.findUndoListByUserId(id).stream().map(todoMapper::createTodoDTOFromEntity).toList();
    }

    @Override
    @Transactional
    public void deleteTodo(long id) {
        todoRepository.deleteById(id);
    }

    @Override
    public void deleteTodoWithStatReward(long id) {
        todoRepository.deleteById(id);
    }
}

File: src/main/java/org/alham/alhamfirst/service/orchestrator/todo/TodoServiceImpl.kt
package org.alham.alhamfirst.service.orchestrator.todo;

import lombok.extern.slf4j.Slf4j;
import org.alham.alhamfirst.common.exception.MariaDBCustomException
import org.alham.alhamfirst.domain.dto.todo.TodoDTO;
import org.alham.alhamfirst.domain.entity.User;
import org.alham.alhamfirst.domain.entity.todo.Todo;
import org.alham.alhamfirst.mapper.TodoMapper;
import org.alham.alhamfirst.repository.todo.TodoRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;


@Service
@Slf4j
@Transactional(readOnly = true)
class TodoServiceImpl(private val todoRepository : TodoRepository
) : TodoService{

    /**
     * 할일 생성
     * startDate는 해당 레이어에서 넣어줌
     *
     * @param todoDTO
     * @return
     */
    @Transactional
    override fun createTodo(todoDTO : TodoDTO) : Todo {
        todoDTO.startDate = LocalDate.now()
        val todo = TodoMapper().createTodoFromDTO(todoDTO)
        val user = User().createTempUser(todoDTO.userId?:0);
        todo.addUser(user)
        return todoRepository.save(todo)
    }

    override fun getTodoList():List<TodoDTO>{
        return todoRepository.findAll().map { TodoMapper().createTodoDTOFromEntity(it) }
    }

    override fun getTodoByIdAndUserId(todoId: Long, userId: Long): Todo {
        return todoRepository.findTodoByIdAndUserId(todoId, userId)?:throw MariaDBCustomException("Todo not found")
    }
    @Transactional
    override fun completeTodo(todoDTO: TodoDTO): TodoDTO {
        val todo = todoRepository.findById(todoDTO.id?:0).orElseThrow{MariaDBCustomException("Todo not found")}
        todo.updateCompeted(todoDTO.completed)
        val saved = todoRepository.save(todo)
        return TodoMapper().createTodoDTOWithUserIdFromEntity(saved)
    }


    override fun getTodoListByUserIdWithUnCompleted(id : Long) : List<TodoDTO>{
        return todoRepository.findUndoListByUserId(id).map{TodoMapper().createTodoDTOFromEntity(it)}
    }
    override fun deleteTodo(id: Long){
        todoRepository.deleteById(id)
    }

    override fun deleteTodoWIthStartReward(id: Long) {
        todoRepository.deleteById(id)
    }


}
